File: scripts/coexpression_heatmap.R
# -*- coding: utf-8 -*-
# 6 个 marker 基因与其余基因的共表达热图 (ComplexHeatmap 版)
#
# 流程与 Python 版一致:
#   1. 读入表达矩阵与 marker 列表
#   2. CPM 标准化后 log2(CPM+1)
#   3. 每个 marker 与其余基因逐一计算 Pearson 相关系数
#   4. 每个 marker 取 |r| 最大的前 TOPN 个基因, 取并集
#   5. 列做层次聚类排序, 输出横版与竖版 PDF, 并导出完整相关系数表
#
# 依赖: readxl, ComplexHeatmap, circlize
#   install.packages(c("readxl", "circlize"))
#   if (!require("BiocManager")) install.packages("BiocManager")
#   BiocManager::install("ComplexHeatmap")

suppressPackageStartupMessages({
  library(readxl)
  library(ComplexHeatmap)
  library(circlize)
  library(grid)
})

## ---------------- 工作目录 ----------------
## 直接把工作目录切到这里, 之后所有相对路径都以它为准:
##   <WORK_ROOT>/data/          输入数据 (xlsx)
##   <WORK_ROOT>/YYYY-MM-DD/    当天的输出, 每天自动新建一个文件夹
WORK_ROOT <- "C:/Users/23027/Desktop/11/画各种图"
dir.create(WORK_ROOT, showWarnings = FALSE, recursive = TRUE)
setwd(WORK_ROOT)
DATA_DIR <- file.path(WORK_ROOT, "data")
dir.create(DATA_DIR, showWarnings = FALSE, recursive = TRUE)
OUT_DIR <- file.path(WORK_ROOT, format(Sys.Date(), "%Y-%m-%d"))
dir.create(OUT_DIR, showWarnings = FALSE, recursive = TRUE)

## ---------------- 输入文件与参数 ----------------
## 输入文件: 同名的 .txt 和 .xlsx 都支持, 优先用存在的那个
pick_file <- function(stem) {
  for (ext in c(".txt", ".tsv", ".csv", ".xlsx")) {
    cand <- file.path(DATA_DIR, paste0(stem, ext))
    if (file.exists(cand)) return(cand)
  }
  file.path(DATA_DIR, paste0(stem, ".xlsx"))
}
EXPR_FILE   <- pick_file("expression_matrix")
MARKER_FILE <- pick_file("marker_genes")
TOPN        <- 20      # 每个 marker 取 |r| 最大的前 N 个基因
FONT        <- "Times New Roman"
GENE_FS     <- 5.5     # 基因名字号
MARKER_FS   <- 8       # marker 名字号
TITLE_FS    <- 10.5    # 标题与图例字号 (五号)

args <- commandArgs(trailingOnly = TRUE)
if (length(args) >= 2) {
  EXPR_FILE <- args[1]
  MARKER_FILE <- args[2]
}

## ---------------- 1. 读入 ----------------
## 按扩展名读取 xlsx 或文本表格 (制表符/逗号分隔)
read_table_any <- function(path, header = TRUE) {
  ext <- tolower(tools::file_ext(path))
  if (ext %in% c("xlsx", "xlsm", "xltx")) {
    as.data.frame(read_excel(path, col_names = header))
  } else {
    sep <- if (ext == "csv") "," else "\t"
    read.table(path, header = header, sep = sep, quote = "",
               comment.char = "", check.names = FALSE,
               stringsAsFactors = FALSE)
  }
}

read_matrix <- function(path) {
  df <- read_table_any(path, header = TRUE)
  df <- df[, colSums(!is.na(df)) > 0, drop = FALSE]   # 去掉全空列
  ids <- as.character(df[[1]])
  m <- as.matrix(df[, -1, drop = FALSE])
  storage.mode(m) <- "double"
  rownames(m) <- ids
  m
}

## 读取 marker 基因名。支持单列基因名文件 (可有可无表头), 或与表达矩阵同结构的文件
read_marker_ids <- function(path) {
  df <- read_table_any(path, header = FALSE)
  ids <- trimws(as.character(df[[1]]))
  ids <- ids[nzchar(ids)]
  if (length(ids) && (startsWith(ids[1], "#") ||
                      tolower(ids[1]) %in% c("id", "gene", "geneid"))) {
    ids <- ids[-1]                                     # 跳过表头
  }
  ids
}

expr <- read_matrix(EXPR_FILE)
marker_ids <- read_marker_ids(MARKER_FILE)
if (!all(marker_ids %in% rownames(expr))) {
  stop("marker 基因不在表达矩阵中: ",
       paste(setdiff(marker_ids, rownames(expr)), collapse = ", "))
}

## ---------------- 2. CPM 标准化 ----------------
cpm <- sweep(expr, 2, colSums(expr), "/") * 1e6
L <- log2(cpm + 1)

other_ids <- setdiff(rownames(L), marker_ids)
Lm <- L[marker_ids, , drop = FALSE]
Lo <- L[other_ids, , drop = FALSE]

## ---------------- 3. 相关系数矩阵 (6 x 其余基因) ----------------
R <- cor(t(Lm), t(Lo), method = "pearson")

## ---------------- 4. 每个 marker 取 topN, 求并集 ----------------
selected <- character(0)
for (i in seq_len(nrow(R))) {
  ord <- order(abs(R[i, ]), decreasing = TRUE)[seq_len(TOPN)]
  selected <- union(selected, colnames(R)[ord])
}
Rsel <- R[, selected, drop = FALSE]

## 列聚类排序 (1 - Pearson 相关距离, average linkage), 两版共用同一顺序
d_genes <- as.dist(1 - cor(Rsel))
hc_genes <- hclust(d_genes, method = "average")

## ---------------- 5. 绘图 ----------------
col_fun <- colorRamp2(c(-1, 0, 1), c("#3C5488", "#FFFFFF", "#E64B35"))
n_gene <- ncol(Rsel)
gene_title <- sprintf("Co-expressed genes (top %d per marker, n = %d)", TOPN, n_gene)
marker_title <- "Marker genes"

legend_param <- list(
  title = "Pearson r",
  at = c(-1, -0.5, 0, 0.5, 1),
  title_gp = gpar(fontsize = TITLE_FS, fontfamily = FONT, fontface = "bold"),
  labels_gp = gpar(fontsize = 9, fontfamily = FONT, fontface = "bold"),
  border = "black"
)

## 横版: marker 作行, 基因作列
ht_h <- Heatmap(
  Rsel,
  name = "Pearson r",
  col = col_fun,
  cluster_rows = FALSE,
  cluster_columns = hc_genes,
  show_column_dend = FALSE,
  row_names_side = "left",
  column_names_side = "bottom",
  row_names_gp = gpar(fontsize = MARKER_FS, fontfamily = FONT, fontface = "bold"),
  column_names_gp = gpar(fontsize = GENE_FS, fontfamily = FONT, fontface = "bold"),
  row_title = marker_title,
  column_title = gene_title,
  column_title_side = "bottom",
  row_title_gp = gpar(fontsize = TITLE_FS, fontfamily = FONT, fontface = "bold"),
  column_title_gp = gpar(fontsize = TITLE_FS, fontfamily = FONT, fontface = "bold"),
  rect_gp = gpar(col = "white", lwd = 0.4),
  border = TRUE,
  heatmap_legend_param = legend_param
)

path_h <- file.path(OUT_DIR, "Fig_coexpression_heatmap_horizontal.pdf")
cairo_pdf(path_h, width = max(6, n_gene * 0.105 + 2.2),
          height = nrow(Rsel) * 0.28 + 2.2, family = FONT)
draw(ht_h, heatmap_legend_side = "right", merge_legend = TRUE)
dev.off()
cat("saved:", normalizePath(path_h), "\n")

## 竖版: 基因作行, marker 作列
Rt <- t(Rsel)
ht_v <- Heatmap(
  Rt,
  name = "Pearson r",
  col = col_fun,
  cluster_rows = hc_genes,
  cluster_columns = FALSE,
  show_row_dend = FALSE,
  row_names_side = "left",
  column_names_side = "bottom",
  row_names_gp = gpar(fontsize = GENE_FS, fontfamily = FONT, fontface = "bold"),
  column_names_gp = gpar(fontsize = MARKER_FS, fontfamily = FONT, fontface = "bold"),
  row_title = gene_title,
  column_title = marker_title,
  column_title_side = "bottom",
  row_title_gp = gpar(fontsize = TITLE_FS, fontfamily = FONT, fontface = "bold"),
  column_title_gp = gpar(fontsize = TITLE_FS, fontfamily = FONT, fontface = "bold"),
  rect_gp = gpar(col = "white", lwd = 0.4),
  border = TRUE,
  heatmap_legend_param = legend_param
)

path_v <- file.path(OUT_DIR, "Fig_coexpression_heatmap_vertical.pdf")
cairo_pdf(path_v, width = ncol(Rt) * 0.28 + 3.0,
          height = max(4, nrow(Rt) * 0.105 + 1.8), family = FONT)
draw(ht_v, heatmap_legend_side = "right", merge_legend = TRUE)
dev.off()
cat("saved:", normalizePath(path_v), "\n")

## ---------------- 6. 导出完整相关系数表 ----------------
## 相关系数 -> t 检验 P 值 (df = n - 2), 再做 BH 多重检验校正
n_sample <- ncol(L)
df_t <- n_sample - 2
tstat <- R * sqrt(df_t) / sqrt(pmax(1 - R^2, 1e-12))
P <- 2 * pt(abs(tstat), df = df_t, lower.tail = FALSE)
Q <- matrix(p.adjust(as.vector(P), method = "BH"), nrow = nrow(P),
            dimnames = dimnames(P))

out_tab <- data.frame(gene = colnames(R), stringsAsFactors = FALSE)
for (i in seq_len(nrow(R))) {
  m <- rownames(R)[i]
  out_tab[[paste0("r_", m)]] <- round(R[i, ], 4)
  out_tab[[paste0("P_", m)]] <- signif(P[i, ], 4)
  out_tab[[paste0("BH_q_", m)]] <- signif(Q[i, ], 4)
}
out_tab$in_heatmap <- ifelse(colnames(R) %in% selected, "yes", "no")
csv_path <- file.path(OUT_DIR, "Fig_coexpression_heatmap_correlation_full.csv")
write.csv(out_tab, csv_path, row.names = FALSE, fileEncoding = "UTF-8")
cat("saved:", normalizePath(csv_path), "\n")

## ---------------- 7. 关键信息 ----------------
cat("\nsamples:", paste(colnames(expr), collapse = ", "), "\n")
cat("library size (raw counts):", paste(colSums(expr), collapse = ", "), "\n")
cat("genes in matrix:", nrow(expr), " markers:", length(marker_ids),
    " others:", length(other_ids), "\n")
cat("genes kept in heatmap:", n_gene, "\n")
cat(sprintf("n = %d samples -> |r| > 0.811 corresponds to P < 0.05 (two-sided)\n",
            n_sample))
cat(sprintf("pairs with P < 0.05: %d / %d\n", sum(P < 0.05), length(P)))
cat(sprintf("pairs with BH q < 0.05: %d / %d\n\n", sum(Q < 0.05), length(Q)))
cat("top 3 co-expressed genes per marker:\n")
for (i in seq_len(nrow(R))) {
  ord <- order(abs(R[i, ]), decreasing = TRUE)[1:3]
  cat(sprintf("  %s: %s\n", rownames(R)[i],
              paste(sprintf("%s (r = %.3f)", colnames(R)[ord], R[i, ord]),
                    collapse = ", ")))
}

File: scripts/coexpression_heatmap.py
# -*- coding: utf-8 -*-
"""
6 个 marker 基因与其余基因的共表达热图。

流程:
  1. 读入表达矩阵 (436 基因 x 6 样本) 与 marker 列表 (6 基因)
  2. CPM 标准化后 log2(CPM+1)  (N 组与 P 组文库大小相差约 4 倍, 必须先标准化)
  3. 每个 marker 与其余基因逐一计算 Pearson 相关系数
  4. 每个 marker 取 |r| 最大的前 TOPN 个基因, 取并集作为热图的列
  5. 对列做层次聚类排序, 输出矢量热图 (PDF + SVG) 与完整相关系数表 (CSV)

用法:
  python coexpression_heatmap.py [表达矩阵文件] [marker文件]   (txt 或 xlsx 均可)
  不带参数时从 <WORK_ROOT>\\data 读取, 结果写入 <WORK_ROOT>\\当天日期 文件夹。
"""

import os
import sys
import datetime
import numpy as np
import openpyxl
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.colors import LinearSegmentedColormap, TwoSlopeNorm

# ---------------- 工作目录 ----------------
# 直接把工作目录切到这里, 之后所有相对路径都以它为准:
#   <WORK_ROOT>\data\          输入数据 (xlsx)
#   <WORK_ROOT>\YYYY-MM-DD\    当天的输出, 每天自动新建一个文件夹
WORK_ROOT = r"C:\Users\23027\Desktop\11\画各种图"
os.makedirs(WORK_ROOT, exist_ok=True)
os.chdir(WORK_ROOT)
DATA_DIR = os.path.join(WORK_ROOT, "data")
os.makedirs(DATA_DIR, exist_ok=True)
OUT_DIR = os.path.join(WORK_ROOT, datetime.date.today().strftime("%Y-%m-%d"))
os.makedirs(OUT_DIR, exist_ok=True)

# 输入文件: 同名的 .txt 和 .xlsx 都支持, 优先用存在的那个
def _pick(stem):
    for ext in (".txt", ".tsv", ".csv", ".xlsx"):
        cand = os.path.join(DATA_DIR, stem + ext)
        if os.path.isfile(cand):
            return cand
    return os.path.join(DATA_DIR, stem + ".xlsx")


EXPR_FILE = _pick("expression_matrix")
MARKER_FILE = _pick("marker_genes")
if len(sys.argv) >= 3:                            # 也可用命令行参数覆盖
    EXPR_FILE, MARKER_FILE = sys.argv[1], sys.argv[2]
OUT_STEM = os.path.join(OUT_DIR, "Fig_coexpression_heatmap")

# ---------------- 参数 ----------------
TOPN = 20              # 每个 marker 取 |r| 最大的前 N 个基因
CORR_METHOD = "pearson"  # 在 log2(CPM+1) 上计算
GENE_LABEL_SIZE = 5.5  # 基因名过多, 单独设小字号; 其余文字为五号 10.5

# ---------------- 字体 ----------------
plt.rcParams["font.family"] = "serif"
plt.rcParams["font.serif"] = ["Times New Roman", "SimSun", "Liberation Serif"]
plt.rcParams["font.sans-serif"] = ["SimSun"]
plt.rcParams["mathtext.fontset"] = "stix"
plt.rcParams["font.weight"] = "bold"
plt.rcParams["axes.labelweight"] = "bold"
plt.rcParams["axes.titleweight"] = "bold"
plt.rcParams["axes.unicode_minus"] = False
plt.rcParams["font.size"] = 10.5
plt.rcParams["axes.labelsize"] = 10.5
plt.rcParams["xtick.labelsize"] = 10.5
plt.rcParams["ytick.labelsize"] = 10.5
plt.rcParams["pdf.fonttype"] = 42
plt.rcParams["ps.fonttype"] = 42
plt.rcParams["svg.fonttype"] = "none"


def _rows_from_file(path):
    """按扩展名读取 xlsx 或 文本表格 (制表符/逗号分隔), 统一返回二维列表"""
    ext = os.path.splitext(path)[1].lower()
    if ext in (".xlsx", ".xlsm", ".xltx"):
        wb = openpyxl.load_workbook(path, read_only=True, data_only=True)
        ws = wb[wb.sheetnames[0]]
        rows = [list(r) for r in ws.iter_rows(values_only=True)]
        wb.close()
        return rows
    with open(path, "r", encoding="utf-8-sig") as fh:
        lines = [ln.rstrip("\r\n") for ln in fh if ln.strip()]
    sep = "\t" if "\t" in lines[0] else ("," if "," in lines[0] else None)
    return [ln.split(sep) if sep else [ln] for ln in lines]


def read_matrix(path):
    """读取表达矩阵, 返回 (样本名, 基因名列表, 计数矩阵)"""
    rows = _rows_from_file(path)
    header = [h for h in rows[0] if h not in (None, "")]
    samples = [str(h) for h in header[1:]]
    genes, values = [], []
    for r in rows[1:]:
        if r[0] in (None, ""):
            continue
        genes.append(str(r[0]).strip())
        values.append([float(r[i + 1]) for i in range(len(samples))])
    return samples, genes, np.array(values, dtype=float)


def read_marker_ids(path):
    """读取 marker 基因名。支持两种格式:
       1) 只有基因名的单列文件 (可有可无表头)
       2) 与表达矩阵同结构的文件, 取第一列"""
    rows = _rows_from_file(path)
    ids = []
    for i, r in enumerate(rows):
        first = str(r[0]).strip() if r and r[0] is not None else ""
        if not first:
            continue
        if i == 0 and (first.startswith("#") or first.lower() in ("id", "gene", "geneid")):
            continue                      # 跳过表头
        ids.append(first)
    return ids


def pearson_vs_matrix(v, M):
    """向量 v 与矩阵 M 每一行的 Pearson r"""
    vc = v - v.mean()
    Mc = M - M.mean(axis=1, keepdims=True)
    denom = np.sqrt((Mc ** 2).sum(axis=1)) * np.sqrt((vc ** 2).sum())
    denom[denom == 0] = np.nan
    return (Mc @ vc) / denom


def order_by_cluster(M):
    """对列做层次聚类排序; 若无 scipy 则退化为按最大相关的 marker 分组排序"""
    try:
        from scipy.cluster.hierarchy import linkage, leaves_list
        from scipy.spatial.distance import pdist
        d = pdist(M.T, metric="correlation")
        d = np.nan_to_num(d, nan=1.0)
        return leaves_list(linkage(d, method="average", optimal_ordering=True))
    except Exception:
        key = np.argmax(np.abs(M), axis=0)
        val = M[key, np.arange(M.shape[1])]
        return np.lexsort((-val, key))


# ---------------- 1. 读入并标准化 ----------------
samples, genes, counts = read_matrix(EXPR_FILE)
marker_ids = read_marker_ids(MARKER_FILE)

cpm = counts / counts.sum(axis=0, keepdims=True) * 1e6
logcpm = np.log2(cpm + 1.0)
gene_index = {g: i for i, g in enumerate(genes)}

missing = [g for g in marker_ids if g not in gene_index]
if missing:
    raise ValueError("marker 基因不在表达矩阵中: {}".format(missing))

other_genes = [g for g in genes if g not in set(marker_ids)]
other_mat = logcpm[[gene_index[g] for g in other_genes]]

# ---------------- 2. 相关系数矩阵 (6 x 其余基因) ----------------
R = np.vstack([pearson_vs_matrix(logcpm[gene_index[m]], other_mat) for m in marker_ids])

# ---------------- 3. 每个 marker 取 topN, 求并集 ----------------
selected = []
for i in range(len(marker_ids)):
    order = np.argsort(-np.abs(np.nan_to_num(R[i])))[:TOPN]
    for j in order:
        if other_genes[j] not in selected:
            selected.append(other_genes[j])
sel_idx = [other_genes.index(g) for g in selected]
Rsel = R[:, sel_idx]
col_order = order_by_cluster(Rsel)
Rsel = Rsel[:, col_order]
sel_names = [selected[j] for j in col_order]

# ---------------- 4. 绘图 ----------------
# 配色: NPG 蓝白红, 蓝为负相关, 红为正相关, 白为 0
# Pearson r 本身的取值范围是 [-1, 1], 故色标固定为 -1 到 1, 不随数据缩放
cmap = LinearSegmentedColormap.from_list(
    "npg_blue_white_red", ["#3C5488", "#FFFFFF", "#E64B35"])
norm = TwoSlopeNorm(vmin=-1.0, vcenter=0.0, vmax=1.0)

n_col = Rsel.shape[1]
n_row = Rsel.shape[0]
GENE_LABEL = "Co-expressed genes (top {} per marker, n = {})".format(TOPN, n_col)
MARKER_LABEL = "Marker genes"


def draw_heatmap(vertical):
    """vertical=False: marker 作行, 基因作列 (横版)
       vertical=True : 基因作行, marker 作列 (竖版)"""
    M = Rsel.T if vertical else Rsel
    nr, nc = M.shape
    if vertical:
        fig_w = n_row * 0.28 + 2.4
        fig_h = max(4.0, n_col * 0.105 + 1.4)
    else:
        fig_w = max(6.0, n_col * 0.105 + 1.8)
        fig_h = n_row * 0.28 + 1.9
    fig, ax = plt.subplots(figsize=(fig_w, fig_h), dpi=300)
    im = ax.imshow(M, cmap=cmap, norm=norm, aspect="auto", interpolation="nearest")

    if vertical:
        ax.set_xticks(np.arange(nc))
        ax.set_xticklabels(marker_ids, rotation=90, fontsize=8, fontweight="bold")
        ax.set_yticks(np.arange(nr))
        ax.set_yticklabels(sel_names, fontsize=GENE_LABEL_SIZE, fontweight="bold")
        ax.set_xlabel(MARKER_LABEL, fontsize=10.5, labelpad=6, fontweight="bold")
        ax.set_ylabel(GENE_LABEL, fontsize=10.5, fontweight="bold")
    else:
        ax.set_xticks(np.arange(nc))
        ax.set_xticklabels(sel_names, rotation=90, fontsize=GENE_LABEL_SIZE, fontweight="bold")
        ax.set_yticks(np.arange(nr))
        ax.set_yticklabels(marker_ids, fontsize=8, fontweight="bold")
        ax.set_xlabel(GENE_LABEL, fontsize=10.5, labelpad=6, fontweight="bold")
        ax.set_ylabel(MARKER_LABEL, fontsize=10.5, fontweight="bold")

    ax.set_xticks(np.arange(-0.5, nc, 1), minor=True)
    ax.set_yticks(np.arange(-0.5, nr, 1), minor=True)
    ax.grid(which="minor", color="white", linewidth=0.4)
    ax.tick_params(which="minor", length=0)
    ax.tick_params(which="major", length=2.0, width=0.8)
    for side in ("left", "bottom", "right", "top"):
        ax.spines[side].set_linewidth(0.8)

    frac = 0.05 if vertical else 0.02
    cbar = fig.colorbar(im, ax=ax, fraction=frac, pad=0.02,
                        ticks=[-1.0, -0.5, 0.0, 0.5, 1.0])
    cbar.set_label("Pearson r", fontsize=10.5, fontweight="bold")
    cbar.ax.tick_params(labelsize=9)
    for lab in cbar.ax.get_yticklabels():
        lab.set_fontweight("bold")
    cbar.outline.set_linewidth(0.8)

    fig.tight_layout()
    stem = OUT_STEM + ("_vertical" if vertical else "_horizontal")
    for ext in ("pdf", "svg", "png"):
        path = "{}.{}".format(stem, ext)
        fig.savefig(path, format=ext, bbox_inches="tight", pad_inches=0.05)
        print("saved:", path)
    plt.close(fig)


draw_heatmap(vertical=False)
draw_heatmap(vertical=True)

# ---------------- 5. 导出完整相关系数表 ----------------
# 相关系数 -> t 检验 P 值 (df = n - 2), 再做 BH 多重检验校正
n_sample = logcpm.shape[1]
df_t = n_sample - 2
with np.errstate(divide="ignore", invalid="ignore"):
    tstat = R * np.sqrt(df_t) / np.sqrt(np.maximum(1.0 - R ** 2, 1e-12))
try:
    from scipy import stats as _stats
    P = 2.0 * _stats.t.sf(np.abs(tstat), df_t)
except Exception:                      # 无 scipy 时用正态近似, 仅作参考
    P = 2.0 * 0.5 * (1.0 - np.math.erf(np.abs(tstat) / np.sqrt(2.0)))
flat = P.ravel()
order = np.argsort(flat)
mtest = flat.size
Q = np.empty(mtest)
running = 1.0
for rank, pos in enumerate(order[::-1]):
    running = min(running, flat[pos] * mtest / (mtest - rank))
    Q[pos] = running
Q = Q.reshape(P.shape)

csv_path = OUT_STEM + "_correlation_full.csv"
with open(csv_path, "w", encoding="utf-8-sig") as fh:
    head = ["gene"]
    for m in marker_ids:
        head += ["r_" + m, "P_" + m, "BH_q_" + m]
    head.append("in_heatmap")
    fh.write(",".join(head) + "\n")
    for j, g in enumerate(other_genes):
        cells = [g]
        for i in range(n_row):
            cells += ["{:.4f}".format(R[i, j]), "{:.4g}".format(P[i, j]),
                      "{:.4g}".format(Q[i, j])]
        cells.append("yes" if g in selected else "no")
        fh.write(",".join(cells) + "\n")
print("saved:", csv_path)

# ---------------- 6. 关键信息输出 ----------------
print("\nsamples:", samples)
print("library size (raw counts):", counts.sum(axis=0).astype(int).tolist())
print("genes in matrix:", len(genes), " markers:", len(marker_ids),
      " others:", len(other_genes))
print("genes kept in heatmap:", n_col)
print("n = {} samples -> |r| > 0.811 corresponds to P < 0.05 (two-sided)".format(n_sample))
print("pairs with P < 0.05: {} / {}".format(int((P < 0.05).sum()), P.size))
print("pairs with BH q < 0.05: {} / {}".format(int((Q < 0.05).sum()), Q.size))
print("\ntop 3 co-expressed genes per marker:")
for i, m in enumerate(marker_ids):
    order = np.argsort(-np.abs(np.nan_to_num(R[i])))[:3]
    txt = ", ".join("{} (r = {:.3f})".format(other_genes[j], R[i, j]) for j in order)
    print("  {}: {}".format(m, txt))
